File: examples/apps/react-native/src/features/store/data/StoreRepositoryImpl.ts
import { ApiClient } from '../../../core/network/ApiClient';
import {
  Business,
  CartItem,
  PaymentResult,
  Product,
  Purchase,
} from '../domain/entities/storeEntities';
import { StoreRepository } from '../domain/repositories/StoreRepository';

export class StoreRepositoryImpl implements StoreRepository {
  constructor(private readonly client: ApiClient) {}

  getProducts() {
    return this.client.get<Product[]>('/api/products');
  }

  getBusinesses() {
    return this.client.get<Business[]>('/api/businesses');
  }

  getBusinessProducts(id: string) {
    return this.client.get<Product[]>(`/api/businesses/${id}/products`);
  }

  async addCartItem(product: Product, quantity: number) {
    await this.client.post('/api/cart/items', {
      productId: product.id,
      quantity,
    });
  }

  async updateCartItem(product: Product, quantity: number) {
    await this.client.patch('/api/cart/items', {
      productId: product.id,
      quantity,
    });
  }

  async removeCartItem(productId: string) {
    await this.client.delete('/api/cart/items', { productId });
  }

  pay(items: CartItem[], total: number) {
    return this.client.post<PaymentResult>('/api/payments', {
      paymentMethod: 'card',
      total,
      items: items.map(item => ({
        productId: item.product.id,
        quantity: item.quantity,
      })),
    });
  }

  getPurchases() {
    return this.client.get<Purchase[]>('/api/purchases');
  }

  getMyProducts() {
    return this.client.get<Product[]>('/api/my-products');
  }

  async createProduct(product: Product) {
    const created = await this.client.post<Product>(
      '/api/my-products',
      product,
    );
    return { ...product, id: created.id };
  }

  async updateProduct(product: Product) {
    await this.client.put(`/api/my-products/${product.id}`, product);
    return product;
  }
}

File: examples/apps/flutter/lib/features/store/data/repositories/store_repository_impl.dart
import 'package:app_prueba_mobilelab/features/store/data/datasources/store_remote_data_source.dart';
import 'package:app_prueba_mobilelab/features/store/domain/entities/business.dart';
import 'package:app_prueba_mobilelab/features/store/domain/entities/cart_item.dart';
import 'package:app_prueba_mobilelab/features/store/domain/entities/product.dart';
import 'package:app_prueba_mobilelab/features/store/domain/entities/purchase.dart';
import 'package:app_prueba_mobilelab/features/store/domain/repositories/store_repository.dart';

class StoreRepositoryImpl implements StoreRepository {
  const StoreRepositoryImpl(this._remote);

  final StoreRemoteDataSource _remote;

  @override
  Future<List<Product>> getProducts() => _remote.getProducts();

  @override
  Future<List<Business>> getBusinesses() => _remote.getBusinesses();

  @override
  Future<List<Product>> getBusinessProducts(String businessId) {
    return _remote.getBusinessProducts(businessId);
  }

  @override
  Future<void> addCartItem(Product product, int quantity) {
    return _remote.addCartItem(product, quantity);
  }

  @override
  Future<void> updateCartItem(Product product, int quantity) {
    return _remote.updateCartItem(product, quantity);
  }

  @override
  Future<void> removeCartItem(String productId) {
    return _remote.removeCartItem(productId);
  }

  @override
  Future<PaymentResult> pay(List<CartItem> items, double total) {
    return _remote.pay(items, total);
  }

  @override
  Future<List<Purchase>> getPurchases() => _remote.getPurchases();

  @override
  Future<List<Product>> getMyProducts() => _remote.getMyProducts();

  @override
  Future<Product> createProduct(Product product) {
    return _remote.createProduct(product);
  }

  @override
  Future<Product> updateProduct(Product product) {
    return _remote.updateProduct(product);
  }
}
